tools/src/main/java/org/apache/kafka/tools/VerifiableConsumer.java

```diff
@@ -0,0 +1,609 @@
+            if (isFinished())
+                break;
+        }
+
+        printJson(new RecordsConsumed(records.count(), summaries));
+        return offsets;
+    }
+
+    @Override
+    public void onComplete(Map<TopicPartition, OffsetAndMetadata> offsets, Exception exception) {
+        List<CommitData> committedOffsets = new ArrayList<>();
+        for (Map.Entry<TopicPartition, OffsetAndMetadata> offsetEntry : offsets.entrySet()) {
+            TopicPartition tp = offsetEntry.getKey();
+            committedOffsets.add(new CommitData(tp.topic(), tp.partition(), offsetEntry.getValue().offset()));
+        }
+
+        boolean success = true;
+        String error = null;
+        if (exception != null) {
+            success = false;
+            error = exception.getMessage();
+        }
+        printJson(new OffsetsCommitted(committedOffsets, error, success));
+    }
+
+    @Override
+    public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+        printJson(new PartitionsAssigned(partitions));
+    }
+
+    @Override
+    public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+        printJson(new PartitionsRevoked(partitions));
+    }
+
+    private void printJson(Object data) {
+        try {
+            out.println(mapper.writeValueAsString(data));
+        } catch (JsonProcessingException e) {
+            out.println("Bad data can't be written as json: " + e.getMessage());
+        }
+    }
+
+    public void commitSync(Map<TopicPartition, OffsetAndMetadata> offsets) {
+        try {
+            consumer.commitSync(offsets);
+            onComplete(offsets, null);
+        } catch (WakeupException e) {
+            // we only call wakeup() once to close the consumer, so this recursion should be safe
+            commitSync(offsets);
+            throw e;
+        } catch (Exception e) {
+            onComplete(offsets, e);
+        }
+    }
+
+    public void run() {
+        try {
+            consumer.subscribe(Arrays.asList(topic), this);
+
+            while (true) {
+                ConsumerRecords<String, String> records = consumer.poll(Long.MAX_VALUE);
+                Map<TopicPartition, OffsetAndMetadata> offsets = onRecordsReceived(records);
+
+                if (!useAutoCommit) {
+                    if (useAsyncCommit)
+                        consumer.commitAsync(offsets, this);
+                    else
+                        commitSync(offsets);
+                }
+            }
+        } catch (WakeupException e) {
+            // ignore, we are closing
+        } finally {
+            consumer.close();
+            shutdownLatch.countDown();
+        }
+    }
+
+    public void close() {
+        boolean interrupted = false;
+        try {
+            consumer.wakeup();
+            while (true) {
+                try {
+                    shutdownLatch.await();
+                    return;
+                } catch (InterruptedException e) {
```

Comment:
I don't think it could, but it seemed like we might as well try to handle the interrupt correctly in case this class was used in another context.